tests/unit/test_space.py

```diff
@@ -438,3 +467,212 @@ def test_box_deepcopy() -> None:
+    with pytest.raises(TF_DEBUGGING_ERROR_TYPES):
+        ProductSearchSpace(spaces=[("A", space_A), ("A", space_B)])
+
+
+def test_product_space_subspace_tags_attribute() -> None:
+    decision_space = Box([-1, -2], [2, 3])
+    context_space = DiscreteSearchSpace(tf.constant([[-0.5, 0.5]]))
+    product_space = ProductSearchSpace(
+        spaces=[("context", context_space), ("decision", decision_space)]
+    )
+
+    npt.assert_array_equal(product_space.subspace_tags, ["context", "decision"])
+
+
+@pytest.mark.parametrize(
+    "spaces, dimension",
+    [
+        ([("A", DiscreteSearchSpace(tf.constant([[-0.5, -0.3], [1.2, 0.4]])))], 2),
+        ([("A", DiscreteSearchSpace(tf.constant([[-0.5]]))), ("B", Box([-1], [2]))], 2),
+        ([("A", Box([-1, -2], [2, 3])), ("B", DiscreteSearchSpace(tf.constant([[-0.5]])))], 3),
+        ([("A", Box([-1, -2], [2, 3])), ("B", Box([-1, -2], [2, 3])), ("C", Box([-1], [2]))], 5),
+    ],
+)
+def test_product_search_space_returns_correct_dimension(
+    spaces: list[tuple[str, SearchSpace]], dimension: int
+) -> None:
+    space = ProductSearchSpace(spaces=spaces)
+    assert space.dimension == dimension
+
+
+def test_product_space_get_subspace() -> None:
+    space_A = Box([-1, -2], [2, 3])
+    space_B = DiscreteSearchSpace(tf.constant([[-0.5, 0.5]]))
+    space_C = Box([-1], [2])
+    product_space = ProductSearchSpace(spaces=[("A", space_A), ("B", space_B), ("C", space_C)])
+
+    subspace_A = product_space.get_subspace("A")
+    assert isinstance(subspace_A, Box)
+    npt.assert_array_equal(subspace_A.lower, [-1, -2])
+    npt.assert_array_equal(subspace_A.upper, [2, 3])
+
+    subspace_B = product_space.get_subspace("B")
+    assert isinstance(subspace_B, DiscreteSearchSpace)
+    npt.assert_array_equal(subspace_B.points, tf.constant([[-0.5, 0.5]]))
+
+    subspace_C = product_space.get_subspace("C")
+    assert isinstance(subspace_C, Box)
+    npt.assert_array_equal(subspace_C.lower, [-1])
+    npt.assert_array_equal(subspace_C.upper, [2])
+
+
+@pytest.mark.parametrize(
+    "spaces, subspace_dim_range",
+    [
+        ([("A", DiscreteSearchSpace(tf.constant([[-0.5]])))], {"A": [0, 1]}),
+        (
+            [
+                ("A", DiscreteSearchSpace(tf.constant([[-0.5]]))),
+                ("B", DiscreteSearchSpace(tf.constant([[-0.5, -0.3], [1.2, 0.4]]))),
+            ],
+            {"A": [0, 1], "B": [1, 3]},
+        ),
+        (
+            [
+                ("A", Box([-1, -2], [2, 3])),
+                ("B", DiscreteSearchSpace(tf.constant([[-0.5]]))),
+                ("C", Box([-1], [2])),
+            ],
+            {"A": [0, 2], "B": [2, 3], "C": [3, 4]},
+        ),
+    ],
+)
+def test_product_space_can_get_subspace_components(
+    spaces: list[tuple[str, SearchSpace]],
+    subspace_dim_range: dict[str, list[int]],
+) -> None:
+    space = ProductSearchSpace(spaces)
+    points = tf.random.uniform([10, space.dimension])
+
+    for tag in space._spaces.keys():
```

Comment:
Accessing private `_space` shall not be necessary. Maybe just iterate over `spaces` instead?